### ISSUE TITLE
C++17 compatibility and ABSEIL

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/obake-feedstock#33]
### Comment:

Hey @bluescarni,

We have been working on ensuring that abseil can be used by many projects at conda-forge.

TLDR, abseil compatibility between libraries is dependent not only on the abseil version, but also on the C++ standard used.

We are considering different alternatives, but the candidates are:

1. Everybody uses the newest abseil and C++17
2. Some packages are not installable together, (at this stage, your package C++20 and many google packages like tensorflow C++17)
3. Some complicated scheme to make multiple ABIs of abseil installable at the same time.

I'm here asking you if:
1. You care that your package is not compatible with other ABSEIL dependencies? If not, maybe option 2 is OK.
2. If you would considering making your package compatible with C++17 and would publish it as such on conda-forge. If so, then we can use option 1.

XREF: The latest discussion is happening https://github.com/conda-forge/abseil-cpp-feedstock/issues/45#issuecomment-1242721630

**Comment by pitrou** [comment]
For the record, according to the Abseil developers this should only be a problem if `obake` uses `absl::bind_front`, at least for now.

**Comment by bluescarni** [comment]
Switching back to C++17 is a no go, as I started making use of several C++20 features (concepts to begin with) and switching back at this point is not realistic.

obake uses only abseil's hashmap, and I have been pretty deliberate in avoiding anything else from abseil. Is it an option to allow the installation of obake with other C++17 projects under the assumption that obake's usage of abseil won't be touching ABI-sensitive areas?

**Comment by pitrou** [comment]
> obake uses only abseil's hashmap, and I have been pretty deliberate in avoiding anything else from abseil. Is it an option to allow the installation of obake with other C++17 projects under the assumption that obake's usage of abseil won't be touching ABI-sensitive areas?

That probably sounds right. @h-vetinari is the best positioned to answer this, though :-)

**Comment by hmaarrfk** [comment]
Thank you for your response. Out of curiosity, do you use any other ABSEIL dependent packages?

**Comment by hmaarrfk** [comment]
in your personal workflows that is.

**Comment by bluescarni** [comment]
> Thank you for your response. Out of curiosity, do you use any other ABSEIL dependent packages?

Not that I know of :) I was under the impression that there aren't many packages depending on abseil currently on conda-forge anyway?

**Comment by hmaarrfk** [comment]
tensorflow and arrow come to mind.

**Comment by pitrou** [comment]
Note that for Arrow it is only an indirect dependency and we don't expose any Abseil types in our APIs.

**Comment by h-vetinari** [comment]
> > obake uses only abseil's hashmap, and I have been pretty deliberate in avoiding anything else from abseil. Is it an option to allow the installation of obake with other C++17 projects under the assumption that obake's usage of abseil won't be touching ABI-sensitive areas?
> 
> That probably sounds right. @h-vetinari is the best positioned to answer this, though :-)

I agree that the risk should be minimal if only using hashmap, but I don't have guarantees for that, as the situation is pretty opaque to me. The current set of control knobs for ABI-relevant differences in upstream abseil seems incomplete, and before we have some more direct confirmation that it is in fact complete(d), we're still nominally in the "only stuff compiled with the same C++ version is compatible" situation.

In other words, this feedstock is fine with the current abseil builds, but it's not a solution that is applicable to the whole ecosystem resp. longterm.
